src/inferencer/constraintStore.ts

```diff
@@ -75,9 +88,28 @@ function solveConstraint(constraintLhs: Type, constraintRhs: Type): any | undefi
     constraintStore.get(constraintLhs) === undefined &&
     constraintStore.get(constraintRhs) !== undefined
   ) {
+    // Rule 7B
+    if (
+      constraintStore.get(constraintLhs) !== undefined &&
+      (constraintStore.get(constraintLhs) as Variable).isAddable &&
+      constraintStore.get(constraintRhs) !== undefined &&
+      isTypeVariable(constraintStore.get(constraintRhs))
+    ) {
+      (constraintStore.get(constraintRhs) as Variable).isAddable = true
+    }
     return solveConstraint(constraintLhs, constraintStore.get(constraintRhs))
   }
-  // Rule 9
+  // Rule 8
+  else if (
+    isFunctionType(constraintLhs) &&
+    constraintStore.get(constraintRhs) !== undefined &&
+    isFunctionType(constraintStore.get(constraintRhs)) &&
+    (constraintLhs as FunctionType).parameterTypes.length ===
+      (constraintStore.get(constraintRhs) as FunctionType).parameterTypes.length
+  ) {
+    addConstraint(constraintLhs as FunctionType, constraintStore.get(constraintRhs))
```

Comment:
This block should do the following:
for each parameter in the constraintLhs, add a constraint for that parameter in Lhs = parameter in Rhs

Then add a constraint from the resultType of constraintLhs = resultType of constraintRhs